### PR TITLE
ports/esp32/modutime.c: add in localtime and mktime

### LINKED ISSUE #202: time.mktime() and time.localtime() are missing
These are a two purely numeric functions that do the calendar transformations between years/months/days to seconds since epoch.  They are in the ESP8266 library:

http://docs.micropython.org/en/latest/esp8266/library/utime.html?highlight=utime
```
MicroPython v1.9.1 on 2017-06-11; ESP module with ESP8266
Type "help()" for more information.
>>> import time
>>> dir(time)
['__name__', 'localtime', 'mktime', 'sleep', 'sleep_ms', 'sleep_us', 'ticks_ms', 'ticks_us', 'ticks_cpu', 'ticks_add', 'ticks_diff', 'time']
>>> 
```

But not the ESP32 library.

```
MicroPython v1.9.2-276-ga9517c04 on 2017-10-09; ESP32 module with ESP32
Type "help()" for more information.
>>> import time
>>> dir(time)
['__name__', 'time', 'sleep', 'sleep_ms', 'sleep_us', 'ticks_ms', 'ticks_us', 'ticks_cpu', 'ticks_add', 'ticks_diff']
>>> 
```

I can't account for how they are missing, as they are mentioned in the utime.rst docs:
    https://github.com/micropython/micropython-esp32/blob/046d15f074398cc2968fc66ae2fd25dc31100109/docs/library/utime.rst

I discovered this because the ESP32 machine module is missing the RTC class, but this is something I can easily work round if I have localtime() and mktime() functions instead.  

Have they been edited out of the utime library accidentally, or put somewhere else?  A more desperate work-around will be to find a pure python implementation of these two functions.

### PR DESCRIPTION
**Pull request by nickzoic**
Uses code copied directly from Josef Gajdusek's contribution to
the esp8266 port modifed to work on esp-idf.

Fixes #202

### CODE REVIEW

File: ports/esp32/modutime.c [hunk 1]
```diff
@@ -29,9 +30,74 @@
 #include <stdio.h>
 #include <string.h>
 #include <sys/time.h>
+#include <py/nlr.h>
+#include <py/smallint.h>
+#include "timeutils.h"
```
Comment (dpgeorge) [inline ports/esp32/modutime.c]: please use the full path: lib/timeutils/timeutils.h

File: ports/esp32/modutime.c [hunk 2]
```diff
@@ -29,9 +30,74 @@
...
+/// contains: (year, month, mday, hour, minute, second, weekday, yearday)
+/// If secs is not provided or None, then the current time from the RTC is used.
+/// year includes the century (for example 2014)
+/// month   is 1-12
+/// mday    is 1-31
+/// hour    is 0-23
+/// minute  is 0-59
+/// second  is 0-59
+/// weekday is 0-6 for Mon-Sun.
+/// yearday is 1-366
```
Comment (dpgeorge) [inline ports/esp32/modutime.c]: No need to copy across these comments.  They are obsolete and the info is provided in the docs.

File: ports/esp32/modutime.c [hunk 3]
```diff
@@ -29,9 +30,74 @@
...
+        tuple[7] = mp_obj_new_int(tm.tm_yday),
+    };
+    return mp_obj_new_tuple(8, tuple);
+}
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(time_localtime_obj, 0, 1, time_localtime);
+
+/// \function mktime()
+/// This is inverse function of localtime. It's argument is a full 8-tuple
+/// which expresses a time as per localtime. It returns an integer which is
+/// the number of seconds since Jan 1, 2000.
```
Comment (dpgeorge) [inline ports/esp32/modutime.c]: ditto

File: ports/esp32/modutime.c [hunk 4]
```diff
@@ -29,9 +30,74 @@
...
+    // localtime generates a tuple of len 8. CPython uses 9, so we accept both.
+    if (len < 8 || len > 9) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "mktime needs a tuple of length 8 or 9 (%d given)", len));
+    }
+
+    return mp_obj_new_int_from_uint(timeutils_mktime(mp_obj_get_int(elem[0]),
+            mp_obj_get_int(elem[1]), mp_obj_get_int(elem[2]), mp_obj_get_int(elem[3]),
+            mp_obj_get_int(elem[4]), mp_obj_get_int(elem[5])));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(time_mktime_obj, time_mktime);
```
Comment (dpgeorge) [inline ports/esp32/modutime.c]: can have "STATIC" at the front of it

File: ports/esp32/modutime.c [hunk 5]
```diff
@@ -29,9 +30,74 @@
...
+        tuple[2] = mp_obj_new_int(tm.tm_mday),
+        tuple[3] = mp_obj_new_int(tm.tm_hour),
+        tuple[4] = mp_obj_new_int(tm.tm_min),
+        tuple[5] = mp_obj_new_int(tm.tm_sec),
+        tuple[6] = mp_obj_new_int(tm.tm_wday),
+        tuple[7] = mp_obj_new_int(tm.tm_yday),
+    };
+    return mp_obj_new_tuple(8, tuple);
+}
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(time_localtime_obj, 0, 1, time_localtime);
```
Comment (dpgeorge) [inline ports/esp32/modutime.c]: can have "STATIC" at the front of it

**Review by dpgeorge — COMMENTED** [review]
Apart from comments it looks good, thanks!

I notice you copied the copyright line from the esp8266 port.  But note that most of the code you copied was copied itself from ports/stm32/modutime.c, so it's arguable that you don't need to add the new copyright line.  But anyway it's not a big deal, please do what you feel is right.

**Comment by dpgeorge** [comment]
Looks good!  Is it ready to merge?

**Comment by nickzoic** [comment]
So far as I'm concerned it's good to go.  I've done some hand testing
and the answers come out right. It's all code copied from elsewhere in the MicroPython project ports
 anyway so no additional copyright issues.

**Comment by dpgeorge** [comment]
Merged in 2305d84844bf77a92d161aae77475a1d1a87101f, thanks!

**Pull request closed without merging**